File: validator/auth/native-config-context.cpp
#include "block/block-auto.h"
#include "block/mc-config.h"
#include "block/transaction.h"

#include "native-config-context.h"
namespace tos::auth {
namespace {
Hash hash(td::Ref<vm::Cell> cell) {
  Hash result{};
  if (cell.not_null()) {
    auto owned = cell->get_hash();
    std::copy_n(owned.as_slice().ubegin(), result.size(), result.begin());
  }
  return result;
}
bool same(td::Ref<vm::Cell> a, td::Ref<vm::Cell> b) {
  return a.is_null() == b.is_null() && (a.is_null() || a->get_hash() == b->get_hash());
}
}  // namespace
Result<ConfigurationAccountData> read_configuration_account(td::Ref<vm::Cell> data) {
  try {
    if (data.is_null())
      return Error{"config-data"};
    vm::CellSlice cs{vm::NoVm{}, std::move(data)};
    // The layout the configuration contract stores: the parameter dictionary,
    // the sequence number and public key, an optional vote dictionary, and the
    // registry checkpoint last. The width is checked before anything is taken
    // from it, so a cell of another shape is refused rather than reinterpreted.
    if (cs.is_special() || cs.size() != 289 || cs.size_refs() < 2)
      return Error{"config-data"};
    ConfigurationAccountData parsed;
    parsed.configuration = cs.fetch_ref();
    if (!cs.advance(288))
      return Error{"config-data"};
    if (cs.fetch_ulong(1) == 1) {
      if (cs.size_refs() != 2)
        return Error{"config-data"};
      cs.fetch_ref();
    }
    if (cs.size_refs() != 1)
      return Error{"config-data"};
    parsed.checkpoint = cs.fetch_ref();
    return parsed;
  } catch (const vm::VmError&) {
    return Error{"config-data"};
  } catch (const vm::VmVirtError&) {
    return Error{"config-data-pruned"};
  }
}

NativeConfigContext::NativeConfigContext(ChainContext chain, Anchor head, Hash address, td::Ref<vm::Cell> code,
                                         td::Ref<vm::Cell> data, td::Ref<vm::Cell> library, td::Ref<vm::Cell> config,
                                         NativeRegistry parent, NativeCommittee committee,
                                         NativeFinalizedHistory history)
    : chain_(chain)
    , head_(head)
    , address_(address)
    , code_(std::move(code))
    , data_(std::move(data))
    , library_(std::move(library))
    , config_(std::move(config))
    , parent_(std::move(parent))
    , committee_(std::move(committee))
    , history_(std::move(history)) {
}
Result<Hash> declared_configuration_account(const block::Config& config, td::Ref<vm::Cell> masterchain_state) {
  auto parameter = config.get_config_param(0);
  if (parameter.is_null())
    return Error{"config-address"};
  vm::CellSlice address_cell{vm::NoVm{}, std::move(parameter)};
  Hash address{};
  if (!address_cell.is_valid() || address_cell.is_special() || address_cell.size() != 256 ||
      address_cell.size_refs() != 0 || !address_cell.fetch_bytes(td::MutableSlice(address.data(), address.size())) ||
      address == Hash{})
    return Error{"config-address"};
  block::gen::ShardStateUnsplit::Record state;
  block::gen::McStateExtra::Record extra;
  block::gen::ConfigParams::Record params;
  if (masterchain_state.is_null() || !tlb::unpack_cell(masterchain_state, state) || state.custom.is_null() ||
      !tlb::unpack_cell(state.custom->prefetch_ref(), extra) || !tlb::csr_unpack(extra.config, params) ||
      params.config_addr != td::Bits256(td::ConstBitPtr(address.data())))
    return Error{"config-address-binding"};
  return address;
}

Result<NativeConfigContext> NativeConfigContext::open(td::Ref<vm::Cell> root, const Anchor& head,
                                                      const ChainContext& chain, const NativeRegistry* cached,
                                                      StateReadBudget budget) {
  try {
    // Establish chain/state trust before extracting account-controlled values.
    auto history = NativeFinalizedHistory::open(root, head, chain, {}, {0, 0});
    if (!history.ok())
      return history.error();
    tos::BlockIdExt id{{tos::masterchainId, tos::shardIdAll, head.seqno_},
                       td::Bits256(td::ConstBitPtr(head.root_.data())),
                       td::Bits256(td::ConstBitPtr(head.file_.data()))};
    auto loaded = block::ConfigInfo::extract_config(
        root, id,
        block::ConfigInfo::needAccountsRoot | block::ConfigInfo::needPrevBlocks | block::Config::needCapabilities);
    if (loaded.is_error())
      return Error{"config-native-state"};
    auto& cfg = *loaded.ok();
    auto declared = declared_configuration_account(cfg, root);
    if (!declared.ok())
      return declared.error();
    const auto address = declared.value();
    block::Account account(-1, td::ConstBitPtr(address.data()));
    auto accounts = cfg.get_accounts_dict();
    if (!account.unpack(accounts.lookup(account.addr), cfg.utime, true) ||
        account.status != block::Account::acc_active || !account.tick || account.addr_rewrite_length != 0 ||
        account.code.is_null() || account.data.is_null() || account.code->get_level() != 0 ||
        account.data->get_level() != 0)
      return Error{"config-account"};
    auto parsed = read_configuration_account(account.data);
    if (!parsed.ok())
      return parsed.error();
    if (!same(parsed.value().configuration, cfg.get_root_cell()))
      return Error{"config-dictionary-binding"};
    auto checkpoint = parsed.value().checkpoint;
    auto registry = cfg.get_config_param(46);
    auto parent = [&]() -> Result<NativeRegistry> {
      if (cached) {
        auto check = cached->checkpoint();
        auto encoded = cached->encode_cell();
        if (!check.ok() || !encoded.ok() || cached->coordinate() != head.seqno_ || !same(check.value(), checkpoint) ||
            !same(encoded.value(), registry))
          return Error{"config-cache-binding"};
        return *cached;
      }
      return NativeRegistry::restore(checkpoint, hash(registry), head.seqno_, budget);
    }();
    if (!parent.ok())
      return parent.error();
    auto committee =
        NativeCommittee::derive(root, head, chain, {tos::masterchainId, tos::shardIdAll}, cfg.cc_seqno_, budget);
    if (!committee.ok())
      return committee.error();
    return NativeConfigContext(chain, head, address, account.code, account.data, account.library, cfg.get_root_cell(),
                               std::move(parent.value()), std::move(committee.value()), std::move(history.value()));
  } catch (const vm::VmError&) {
    return Error{"config-native-state"};
  } catch (const vm::VmVirtError&) {
    return Error{"config-pruned"};
  }
}
Result<bool> NativeConfigContext::binds(std::int32_t workchain, const Hash& address, td::Ref<vm::Cell> code,
                                        td::Ref<vm::Cell> data, td::Ref<vm::Cell> library) const {
  if (workchain != -1 || address != address_ || !same(code, code_) || !same(data, data_) || !same(library, library_))
    return Error{"config-transaction-binding"};
  return true;
}
}  // namespace tos::auth

File: tosctl/src/validator-auth-native/src/native_config_context.rs
use crate::{
    committee::NativeCommittee,
    native,
    native_apply::CurrentRegistry,
    native_history::{HistoryReadBudget, NativeFinalizedHistory},
    native_registry::NativeRegistry,
    registry::StateReadBudget,
};
use chain_block::{
    BlockIdExt, Cell, CellType, Deserializable, HashmapType, McStateExtra, Serializable,
    ShardStateUnsplit, SliceData,
};
use tos_validator_auth::{
    codec::{Error, Hash},
    context::ChainContext,
    types::Anchor,
};
type OfflineHistory = NativeFinalizedHistory<fn(&BlockIdExt, usize) -> Result<Vec<u8>, Error>>;
fn unavailable(_: &BlockIdExt, _: usize) -> Result<Vec<u8>, Error> {
    Err(Error("history-block-unavailable"))
}
fn parameter(config: &chain_block::ConfigParams, index: u32) -> Result<Cell, Error> {
    let entry = native(config.config_params.get(native(index.write_to_bitstring())?))?
        .ok_or(Error("native-config"))?;
    if entry.remaining_bits() != 0 || entry.remaining_references() != 1 {
        return Err(Error("native-config"));
    }
    native(entry.reference(0))
}
fn same(a: &Option<Cell>, b: &Option<Cell>) -> bool {
    match (a, b) {
        (Some(a), Some(b)) => a.repr_hash() == b.repr_hash(),
        (None, None) => true,
        _ => false,
    }
}
/// Immutable native parent authority. Account bytes are selected from the actual
/// authenticated ShardAccounts dictionary, never supplied by a contract or C7.
pub struct NativeConfigContext {
    chain: ChainContext,
    head: Anchor,
    address: Hash,
    code: Cell,
    data: Cell,
    library: Option<Cell>,
    config: Cell,
    parent: NativeRegistry,
    committee: NativeCommittee,
    history: OfflineHistory,
}
impl NativeConfigContext {
    pub fn open(
        root: Cell,
        head: &Anchor,
        chain: &ChainContext,
        cached: Option<&NativeRegistry>,
        budget: StateReadBudget,
    ) -> Result<Self, Error> {
        let history = NativeFinalizedHistory::open(
            root.clone(),
            head.clone(),
            chain.clone(),
            unavailable as fn(&BlockIdExt, usize) -> Result<Vec<u8>, Error>,
            HistoryReadBudget { blocks: 0, bytes: 0 },
        )?;
        let state = native(ShardStateUnsplit::construct_from_full_cell(root.clone()))?;
        let extra = native(McStateExtra::construct_from_full_cell(
            state.custom_cell().ok_or(Error("config-native-state"))?,
        ))?;
        let config = &extra.config;
        let address_cell = parameter(config, 0)?;
        if address_cell.cell_type() != CellType::Ordinary
            || address_cell.bit_length() != 256
            || address_cell.references_count() != 0
        {
            return Err(Error("config-address"));
        }
        let mut a = native(SliceData::load_cell(address_cell))?;
        let address: Hash =
            native(a.get_next_bits(256))?.try_into().map_err(|_| Error("config-address"))?;
        if address == [0; 32] {
            return Err(Error("config-address"));
        }
        if config.config_addr.remaining_bits() != 256
            || config.config_addr.get_bytestring(0) != address
        {
            return Err(Error("config-address-binding"));
        }
        let accounts = native(state.read_accounts())?;
        let shard_account =
            native(accounts.account(&config.config_addr))?.ok_or(Error("config-account"))?;
        let account = native(shard_account.read_account())?;
        let addr = account.get_addr().ok_or(Error("config-account"))?;
        let init = account.state_init().ok_or(Error("config-account"))?;
        let code = account.get_code().ok_or(Error("config-account"))?;
        let data = account.get_data().ok_or(Error("config-account"))?;
        if addr.workchain_id() != -1
            || addr.rewrite_pfx().is_some()
            || addr.address().get_bytestring(0) != address
            || account.get_tick_tock().is_none_or(|t| !t.tick)
            || code.level() != 0
            || data.level() != 0
            || account.last_tr_time().ok_or(Error("config-account"))?.max(1)
                <= shard_account.last_trans_lt()
        {
            return Err(Error("config-account"));
        }
        let mut d = native(SliceData::load_cell(data.clone()))?;
        if d.cell_type() != CellType::Ordinary
            || d.remaining_bits() != 289
            || d.remaining_references() < 2
        {
            return Err(Error("config-data"));
        }
        let owned_config = native(d.checked_drain_reference())?;
        let actual_config = config.config_params.data().ok_or(Error("config-native-state"))?;
        if owned_config.repr_hash() != actual_config.repr_hash() {
            return Err(Error("config-dictionary-binding"));
        }
        native(d.get_next_bits(288))?;
        if native(d.get_next_bit())? {
            if d.remaining_references() != 2 {
                return Err(Error("config-data"));
            }
            native(d.checked_drain_reference())?;
        }
        if d.remaining_references() != 1 {
            return Err(Error("config-data"));
        }
        let checkpoint = native(d.checked_drain_reference())?;
        let registry = parameter(config, 46)?;
        let parent = match cached {
            Some(value) => {
                if value.coordinate() != head.seqno
                    || value.checkpoint()?.repr_hash() != checkpoint.repr_hash()
                    || value.encode_cell()?.repr_hash() != registry.repr_hash()
                {
                    return Err(Error("config-cache-binding"));
                }
                value.clone()
            }
            None => NativeRegistry::restore(
                checkpoint,
                registry.repr_hash().as_slice(),
                head.seqno,
                budget,
            )?,
        };
        let committee = NativeCommittee::derive(
            root,
            head,
            chain,
            -1,
            1 << 63,
            extra.validator_info.catchain_seqno,
            budget,
        )?;
        Ok(Self {
            chain: chain.clone(),
            head: head.clone(),
            address,
            code,
            data,
            library: init.library.root().cloned(),
            config: actual_config.clone(),
            parent,
            committee,
            history,
        })
    }
    pub fn binds(
        &self,
        workchain: i32,
        address: &Hash,
        code: Cell,
        data: Cell,
        library: Option<Cell>,
    ) -> Result<bool, Error> {
        if workchain != -1
            || address != &self.address
            || code.repr_hash() != self.code.repr_hash()
            || data.repr_hash() != self.data.repr_hash()
            || !same(&library, &self.library)
        {
            return Err(Error("config-transaction-binding"));
        }
        Ok(true)
    }
    pub fn chain(&self) -> &ChainContext {
        &self.chain
    }
    pub fn head(&self) -> &Anchor {
        &self.head
    }
    pub fn address(&self) -> &Hash {
        &self.address
    }
    pub fn parent(&self) -> &NativeRegistry {
        &self.parent
    }
    pub fn committee(&self) -> &NativeCommittee {
        &self.committee
    }
    pub fn history(&self) -> &OfflineHistory {
        &self.history
    }
    pub fn config(&self) -> Cell {
        self.config.clone()
    }
    pub fn data(&self) -> Cell {
        self.data.clone()
    }
}
